src/binary_basic_decode.ml

```diff
@@ -0,0 +1,639 @@
+  (c, next_input)
+
+(* https://en.wikipedia.org/wiki/LEB128#Unsigned_LEB128 *)
+let rec read_UN n input =
+  assert (n > 0);
+  let* b, input = read_byte input in
+  let b = Char.code b in
+  let x = Int64.of_int (b land 0x7f) in
+  if b land 0x80 = 0 then Ok (x, input)
+  else
+    let+ i64, input = read_UN (n - 7) input in
+    (Int64.logor x (Int64.shl i64 7L), input)
+
+let read_U32 input =
+  let+ i64, input = read_UN 32 input in
+  (Int64.to_int i64, input)
+
+let read_U64 input =
+  let+ i64, input = read_UN 64 input in
+  (Int64.to_int i64, input)
+
+(* https://en.wikipedia.org/wiki/LEB128#Signed_LEB128 *)
+let rec read_SN n input =
+  assert (n > 0);
+  let* b, input = read_byte input in
+  let b = Char.code b in
+  let x = Int64.of_int (b land 0x7f) in
+  if b land 0x80 = 0 then
+    if b land 0x40 = 0 then Ok (x, input)
+    else Ok (Int64.(logor x (logxor (-1L) 0x7fL)), input)
+  else
+    let* i64, input = read_SN (n - 7) input in
+    Ok (Int64.logor x (Int64.shl i64 7L), input)
+
+let read_S32 input =
+  let+ i64, input = read_SN 32 input in
+  (Int64.to_int32 i64, input)
+
+let read_S64 input =
+  let+ i64, input = read_SN 64 input in
+  (i64, input)
+
+let read_F32 input =
+  let i32_of_byte input =
+    let+ b, input = read_byte input in
+    (Int32.of_int (int_of_char b), input)
+  in
+  let* i1, input = i32_of_byte input in
+  let* i2, input = i32_of_byte input in
+  let* i3, input = i32_of_byte input in
+  let+ i4, input = i32_of_byte input in
+  let i32 = Int32.shl i4 24l in
+  let i32 = Int32.logor i32 (Int32.shl i3 16l) in
+  let i32 = Int32.logor i32 (Int32.shl i2 8l) in
+  let i32 = Int32.logor i32 i1 in
+  (Float32.of_bits i32, input)
+
+let read_F64 input =
+  let i64_of_byte input =
+    let+ b, input = read_byte input in
+    (Int64.of_int (int_of_char b), input)
+  in
+  let* i1, input = i64_of_byte input in
+  let* i2, input = i64_of_byte input in
+  let* i3, input = i64_of_byte input in
+  let* i4, input = i64_of_byte input in
+  let* i5, input = i64_of_byte input in
+  let* i6, input = i64_of_byte input in
+  let* i7, input = i64_of_byte input in
+  let+ i8, input = i64_of_byte input in
+  let i64 = Int64.shl i8 56L in
+  let i64 = Int64.logor i64 (Int64.shl i7 48L) in
+  let i64 = Int64.logor i64 (Int64.shl i6 40L) in
+  let i64 = Int64.logor i64 (Int64.shl i5 32L) in
+  let i64 = Int64.logor i64 (Int64.shl i4 24L) in
+  let i64 = Int64.logor i64 (Int64.shl i3 16L) in
+  let i64 = Int64.logor i64 (Int64.shl i2 8L) in
+  let i64 = Int64.logor i64 i1 in
+  (Float64.of_bits i64, input)
+
+let vector parse_elt input =
+  let* nb_elt, input = read_U32 input in
+  let rec loop loop_id input acc =
+    if nb_elt = loop_id then Ok (List.rev acc, input)
+    else
+      let* acc_elt, input = parse_elt loop_id input in
+      let acc = acc_elt :: acc in
+      loop (loop_id + 1) input acc
+  in
+  loop 0 input []
+
+let vector_no_id f input = vector (fun _id -> f) input
+
+let deserialize_indice input =
```

Comment:
If you make the type more precise by using something along the lines of:

```ocaml
let deserialize_indice input : Types.simplified Types.indice =
  ...
```

Then it should simplify some other parts of the code such as:

```ocaml
let tableidx =
  match tableidx with Raw i -> i | _ -> assert false
```

Because the typechecker will understand that the other case is impossible so you'll be able to simply write:

```ocaml
let tableidx =
  match tableidx with Raw i -> i
```

Or even:

```ocaml
let (Raw tableidx) = tableidx in
```

Reply:
Complete type with `result`
```ocaml
(Types.simplified Types.indice * Input.t, Error.err) result
```